### DISCUSSION TITLE
Integration of SVG Rasterization

### DISCUSSION THREAD

**Discussion by Michael5601** [eclipse-platform/eclipse.platform.swt#1651] [General]
Right now the proposed [feature](https://github.com/eclipse-platform/eclipse.platform.swt/pull/1638) for SVG rasterization utilizes a new public API type `SVGRasterizer` (former `ISVGRasterizer`). The functionality is provided by the class `JSVGRasterizer` which implements the interface. 

We had a discussion [here](https://github.com/eclipse-platform/eclipse.platform.swt/pull/1638#issuecomment-2525188884) about wether this public API should be replaced by an internal class as SVG support will be integrated in SWT in the long run. 

On the other hand the public API allows for low coupling and flexibility if we want to exchange the functionality of the rasterizer without rebuilding SWT. @HeikoKlare also mentioned [here](https://github.com/eclipse-platform/eclipse.platform.swt/pull/1638#issuecomment-2530867206) that the interface could be internal for now as a compromise which would allow integration of a different rasterizer while avoiding the public visibility.

**Comment by HeikoKlare** [comment]
I am in favor of making the new API internal (and keep it internal unless there is some requirement for making it public).

Maybe my initial comments on this concern were misleading: I just proposed to separate the interface from the actual service implementation for architectural reasons. It allows to still use SWT without SVG support (and in particular according dependencies), ensured by proper encapsulation in separate bundles rather than just by trying to avoid "invalid"/unnecessary dependencies within a single bundle. And it ensure that one _may_ exchange the rasterizer, e.g., if someone really needs it (for performance reasons or thel ike) or, more likely, if, at some point in time, JSVG may become outdated and we need to implement a new rasterizer based on a different library. Then we can easily exchange the fragment for development, testing, CI, deployment etc. purposes, while a tight integration into the SWT host requires a complete exchange of the host bundle.

But from a pure consumer perspective, we (= Vector Informatik) currently have no need of making the rasterizer exchangeable.

**Reply by Michael5601** [reply]
I performed the change now by moving `SVGRasterizerRegistry` and `SVGRasterizer` into `org.eclipse.swt.internal`. I will close the discussion as we have a consensus.

**Comment by selundqma** [comment]
My team has looked forward to using the JSVGRasterizer class to avoid having a lot of PNG images in our applications. Last version of Eclipse we could use was v4.34 (2024-12) and now we REALLY wanted to make use of the new rasterizer in v4.38 (2025-12).. But.. what do we see?? Discouraged access?

<img width="504" height="144" alt="image" src="https://github.com/user-attachments/assets/2eea1abe-de4a-4a71-a6c8-92589326afd3" />

Should we use the com.github.weisj.jsvg classes "directly" and avoid using the org.eclipse.swt.svg.JSVGRasterizer at all? Although it does not seem to be tagged for internal use by the name of the package?

**Reply by HeikoKlare** [reply]
I am not sure what you are trying to do with the `JSVGRasterizer` class. The `JSVGRasterizer` is the currently used implementation for rasterizing SVGs. But that's an implementation detail of SWT which you should not deal with at all. You can simply use load SVGs in the common ways, such as:
- Create an `Image` with a file path or `ImageFileNameProvider` using a path to an SVG or an `InputStream` containing an SVG
- Use JFace image descriptors with paths/URLs to SVG files

**Reply by laeubi** [reply]
As Heiko correctly explained you can already use the SVG image support in SWT and its working quite well from my first test also for my own SVGs. The only limitation currently is that you can not (without access internal API) get an image at a given size. This is something currently worked on in https://github.com/eclipse-platform/eclipse.platform.ui/pull/3431 currently.

**Reply by selundqma** [reply]
Thanks for your quick replies!!! Love it!! Yeah, we really need to get images at a given size. I will keep track of the progress of the issue you mentioned. Thanks!

**Reply by HeikoKlare** [reply]
If you have specific use cases for getting images at a specified size, it would be great if you can share them, so that for any newly introduced concept/API we can take them into account.
Note that if you just want to draw an image/SVG at a specified size on a GC in a sharp way, this is already possible: https://eclipse.dev/eclipse/markdown/?f=news/4.38/platform_isv.md#new-gc-api-for-drawing-scaled-images

**Reply by selundqma** [reply]
Use case is to make use of the setImage method in the Label class to show an image.

**Reply by HeikoKlare** [reply]
So you would like to create an SVG and load/rasterize that at different sizes to set as an image for different labels?

**Reply by selundqma** [reply]
Yes, I have some SVGs and each of needs to be scaled to different image sizes and set as image for different labels. Like, when you hover over an "image button" you get a larger image.. To see which image the mouse pointer is over.

**Reply by HeikoKlare** [reply]
I see, thank you for the information. Definitely sounds like a use case for the mentioned https://github.com/eclipse-platform/eclipse.platform.ui/pull/3431
